pkg/scalers/openstack/keystone_authentication.go

```diff
@@ -0,0 +1,206 @@
+	User *userProps `json:"user"`
+}
+
+type appCredentialProps struct {
+	ID     string `json:"id"`
+	Secret string `json:"secret"`
+}
+
+type scopeProps struct {
+	Project *projectProps `json:"project"`
+}
+
+type userProps struct {
+	ID       string `json:"id"`
+	Password string `json:"password"`
+}
+
+type projectProps struct {
+	ID string `json:"id"`
+}
+
+// GetToken retrieves a token from Keystone
+func (authProps *KeystoneAuthMetadata) GetToken() (string, error) {
+	jsonBody, jsonError := json.Marshal(authProps)
+
+	if jsonError != nil {
+		return "", jsonError
+	}
+
+	body := bytes.NewReader(jsonBody)
+
+	tokenURL, err := url.Parse(authProps.AuthURL)
+
+	if err != nil {
+		return "", fmt.Errorf("the authURL is invalid: %s", err.Error())
+	}
+
+	tokenURL.Path = path.Join(tokenURL.Path, tokensEndpoint)
+
+	getTokenRequest, getTokenRequestError := http.NewRequest("POST", tokenURL.String(), body)
+
+	getTokenRequest.Header.Set("Content-Type", "application/json")
+
+	if getTokenRequestError != nil {
+		return "", getTokenRequestError
+	}
+
+	resp, requestError := authProps.HTTPClient.Do(getTokenRequest)
+
+	if requestError != nil {
+		return "", requestError
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+		authProps.AuthToken = resp.Header["X-Subject-Token"][0]
+		return resp.Header["X-Subject-Token"][0], nil
+	}
+
+	errBody, _ := ioutil.ReadAll(resp.Body)
+
+	return "", fmt.Errorf(string(errBody))
+}
+
+// IsTokenValid checks if a authentication token is valid
+func IsTokenValid(authProps KeystoneAuthMetadata) (bool, error) {
+	token := authProps.AuthToken
+
+	tokenURL, err := url.Parse(authProps.AuthURL)
+
+	if err != nil {
+		return false, fmt.Errorf("the authURL is invalid: %s", err.Error())
+	}
+
+	tokenURL.Path = path.Join(tokenURL.Path, tokensEndpoint)
+
+	checkTokenRequest, checkRequestError := http.NewRequest("HEAD", tokenURL.String(), nil)
+	checkTokenRequest.Header.Set("X-Subject-Token", token)
+	checkTokenRequest.Header.Set("X-Auth-Token", token)
+
+	if checkRequestError != nil {
+		return false, checkRequestError
+	}
+
+	checkResp, requestError := authProps.HTTPClient.Do(checkTokenRequest)
+
+	if requestError != nil {
+		return false, requestError
+	}
+
+	defer checkResp.Body.Close()
+
+	if checkResp.StatusCode >= 400 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// NewPasswordAuth creates a struct containing metadata for authentication using password method
+func NewPasswordAuth(authURL string, userID string, userPassword string, projectID string, httpTimeout int) (*KeystoneAuthMetadata, error) {
+	var tokenError error
+
+	passAuth := new(KeystoneAuthMetadata)
+
+	passAuth.Properties = new(authProps)
+
+	passAuth.Properties.Scope = new(scopeProps)
+	passAuth.Properties.Scope.Project = new(projectProps)
+
+	passAuth.Properties.Identity = new(identityProps)
+	passAuth.Properties.Identity.Password = new(passwordProps)
+	passAuth.Properties.Identity.Password.User = new(userProps)
```

Comment:
it seems to me that you don't need to make `passAuth`, `passAuth.Properties`, `passAuth.Properties.Identity`, `passAuth.Properties.Identity.Password`, `passAuth.Properties.Identity.Password.User`, or `passAuth.Properties.Scope` references, and then you'll get a default object for all of them just by doing 

```go
passAuth := KeystoneAuthMetadata{}
```

right?

Reply:
As far as I know (and tried), since `KeystoneAuthMetadata` is a struct that has other structs inside that are pointers, they need to be initialized somehow first. Thus, 

```go
passAuth := KeystoneAuthMetadata{}
```

is not sufficient to get a proper `KeystoneAuthMetadata` default object. Assigning values to it without manually making those references will result in: 
```
runtime error: invalid memory address or nil pointer dereference
```

Maybe there is a more elegant way to do or another way to initialize these "inner structs", but my limited experience in Go does not allow me to find it out 😅

Reply:
I guess what I meant is maybe non of these pointers need to be a pointer since you never expect them to actually be nil, right? So `KeystoreAuthMetadata` would be 

```diff
type KeystoneAuthMetadata struct {
	AuthURL    string       `json:"-"`
	AuthToken  string       `json:"-"`
	HTTPClient *http.Client `json:"-"`
-	Properties *authProps   `json:"auth"`
+	Properties authProps   `json:"auth"`
}
```

and in turn `authProps` would be 

```diff
type authProps struct {
-	Identity *identityProps `json:"identity"`
+	Identity identityProps `json:"identity"`
-	Scope    *scopeProps    `json:"scope,omitempty"`
+	Scope    scopeProps    `json:"scope,omitempty"`
}
```

and so on. 

I think the only difference is in the `json.Marshal()` behavior where a nil is an empty object but any non-nil object is not empty, but I don't think it makes a difference really. Just thought it would make the code a bit easier to read since you're already initializing all the pointers with new() anyway. No worries if it gives you trouble.

Reply:
Exactly! They do not need to be pointers. The only reason we did that way was because of `json.Marshal()` as you noticed. Since we use these structs to build a JSON payload for our token request, some of them need to be omitted (and thus be pointers, and then be `nil`) in order to make it work properly.

Also, we plan to submit new scalers for OpenStack projects soon. As they come, we can study better ways to do it and, eventually, updated the way we originally structured this.

That is nice feedback, anyway @ahmelsayed. Thanks! We'll keep an eye on it 😄